src/XtensaOptimize.cpp

```diff
@@ -0,0 +1,2331 @@
+            ramp(wild_i32, 1, pred.type().lanes()) <= bc(wild_i32, pred.type().lanes())};
+
+        vector<Expr> matches;
+        Expr new_pred;
+        for (const Expr &p : patterns) {
+            if (expr_match(p, pred, matches)) {
+                for (auto &m : matches) {
+                    m = mutate(m);
+                }
+                new_pred = Call::make(pred.type(), "clamped_dense_ramp", matches, Call::PureExtern);
+                break;
+            }
+        }
+        return new_pred;
+    }
+
+    Expr visit(const Load *op) override {
+        if (!is_const_one(op->predicate)) {
+            Expr new_pred = match_clamped_dense_ramp(op->index, op->predicate);
+
+            if (new_pred.defined()) {
+                return Load::make(op->type, op->name,
+                                  mutate(op->index), op->image,
+                                  op->param,
+                                  new_pred,
+                                  op->alignment);
+            }
+        }
+
+        return IRGraphMutator::visit(op);
+    }
+
+    Stmt visit(const Store *op) override {
+        if (!is_const_one(op->predicate)) {
+            Expr new_pred = match_clamped_dense_ramp(op->index, op->predicate);
+
+            if (new_pred.defined()) {
+                return Store::make(op->name, mutate(op->value), mutate(op->index),
+                                   op->param, new_pred, op->alignment);
+            }
+        }
+
+        return IRGraphMutator::visit(op);
+    }
+
+public:
+    MatchXtensaPatterns(const Target &target)
+        : target(target) {
+    }
+};
+
+// Find an upper bound of bounds.max - bounds.min.
+Expr span_of_bounds(const Interval &bounds) {
+    internal_assert(bounds.is_bounded());
+
+    const Min *min_min = bounds.min.as<Min>();
+    const Max *min_max = bounds.min.as<Max>();
+    const Min *max_min = bounds.max.as<Min>();
+    const Max *max_max = bounds.max.as<Max>();
+    const Add *min_add = bounds.min.as<Add>();
+    const Add *max_add = bounds.max.as<Add>();
+    const Sub *min_sub = bounds.min.as<Sub>();
+    const Sub *max_sub = bounds.max.as<Sub>();
+
+    if (min_min && max_min && equal(min_min->b, max_min->b)) {
+        return span_of_bounds({min_min->a, max_min->a});
+    } else if (min_max && max_max && equal(min_max->b, max_max->b)) {
+        return span_of_bounds({min_max->a, max_max->a});
+    } else if (min_add && max_add && equal(min_add->b, max_add->b)) {
+        return span_of_bounds({min_add->a, max_add->a});
+    } else if (min_sub && max_sub && equal(min_sub->b, max_sub->b)) {
+        return span_of_bounds({min_sub->a, max_sub->a});
+    } else {
+        return bounds.max - bounds.min;
+    }
+}
+
+// NOTE(vksnk): this is borrowed from HexagonOptimize.cpp, so
+// eventually need to generalize and share across two places.
+// Replace indirect loads with dynamic_shuffle intrinsics where
+// possible.
+class OptimizeShuffles : public IRMutator {
```

Comment:
If this is directly copied without change, it should probably be moved into a separate file and de-duplicated.